app/controllers/etds_controller.rb

```diff
@@ -0,0 +1,163 @@
+  end
+
+  def reader_action_attributes
+    {
+      readerapproval: etd_params[:readerapproval],
+      readercomment: etd_params[:readercomment],
+      last_reader_action_at: etd_params[:readeractiondttm]&.in_time_zone(Settings.peoplesoft_timezone)
+    }
+  end
+
+  def registrar_action_attributes
+    {
+      regapproval: etd_params[:regapproval],
+      regcomment: etd_params[:regcomment],
+      last_registrar_action_at: etd_params[:regactiondttm]&.in_time_zone(Settings.peoplesoft_timezone)
+    }
+  end
+
+  def render_bad_request
+    render status: :bad_request, html: invalid_xml_message
+  end
+
+  def render_created
+    render status: :created, html: "#{submission.druid} #{message}"
+  end
+
+  # Authentication based on a allow list of server names and IP addresses
+  def request_from_authorized_origin?
+    begin
+      remote_hostname = Resolv.getname(request.remote_addr).downcase
+      return true if Settings.ps_servers.include?(remote_hostname)
+    rescue Resolv::ResolvError
+      return true if Settings.ps_ips.include?(request.remote_addr)
+    end
+    false
+  end
+
+  def authenticate
+    http_basic_authenticate_or_request_with(
+      name: Settings.dlss_admin,
+      password: Settings.dlss_admin_pw,
+      realm: 'Application',
+      message: 'You are unauthorized to perform this action'
+    )
+  end
+
+  def set_submission
+    render_bad_request and return unless valid_xml?
+
+    @submission = Submission.find_or_initialize_by(dissertation_id:, title:)
+    @message = @submission.new_record? ? 'created' : 'updated'
+  end
+
+  # Translate the incoming XML data to attributes on the Submission model
+  # Exclude reader and registrar actions which are handled separately
+  def submission_attributes # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+    # Reader and Registratar actions are handled separately
+    # so remove them from the attributes to be updated here
+    etd_params.to_h.except(:dissertationid, :reader, :title,
+                           :readerapproval, :readercomment, :readeractiondttm,
+                           :regapproval, :regcomment, :regactiondttm).tap do |attrs|
+      attrs[:dissertation_id] = dissertation_id
+      attrs[:title] = title
+      attrs[:etd_type] = attrs.delete(:type)
+      attrs[:ps_career] = attrs.delete(:career)
+      attrs[:ps_program] = attrs[:program]
+      attrs[:department] = attrs.delete(:program)
+      attrs[:ps_plan] = attrs[:plan]
+      attrs[:major] = attrs.delete(:plan)
+      attrs[:degree] = degree
+      attrs[:ps_subplan] = attrs.delete(:subplan)[:__content__] if attrs[:subplan]
+      attrs[:sub] = "deadline #{attrs.delete(:sub)[:deadline]}" if attrs[:sub]
+      attrs[:provost] = attrs.delete(:vpname)
+      attrs[:name] = attrs.delete(:name)&.gsub(/,(\S)/, ', \1')
+    end.compact
+  end
+
+  def valid_xml?
+    true if etd_params && dissertation_id && readers
+  rescue StandardError => e
```

Comment:
One change here is that the current app will return an `:internal_server_error` if the xml can't be parsed, the rest are `:bad_request`. I could capture different errors here and try to maintain that if it's critical.